drivers: sensor: lsm6dsl: add SPI support

avisconti opened the pull request:
Add SPI bus support to LSM6DSL sensor. The bus routines (i.e. I2C
and SPI) are defined in separate files, where proper r/w callbacks
are registered. SPI support in driver is currently not providing
the capability of using gpios as CS.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

avisconti committed:
drivers: sensor: lsm6dsl: add SPI support

Add SPI bus support to LSM6DSL sensor. The bus routines (i.e. I2C
and SPI) are defined in separate files, where proper r/w callbacks
are registered.

Signed-off-by: Armando Visconti <armando.visconti@st.com>

tbursztyka commented on drivers/sensor/lsm6dsl/lsm6dsl.h:
```diff
@@ -689,12 +688,20 @@
+struct lsm6dsl_transfer_function {
+	int (*read_data)(struct lsm6dsl_data *data, uint8_t reg_addr, uint8_t *value, uint8_t len);
+	int (*read_reg)(struct lsm6dsl_data *data, uint8_t reg_addr, uint8_t *value);
+	int (*update_reg)(struct lsm6dsl_data *data, uint8_t reg_addr, uint8_t value, uint8_t mask);
```
beware of the 80 chars limit 

apply the relevant change everywhere where relevant.

avisconti replied:
ack

tbursztyka commented on drivers/sensor/lsm6dsl/lsm6dsl_spi.c:
```diff
@@ -0,0 +1,106 @@
+#define SPI_CS NULL
+
+static struct spi_config lsm6dsl_spi_conf = {
+	.frequency = 16000000,
```
you'll need to make it Kconfig based, because if the spi master controller device is not able to go that high (it's not high, but some controllers are twice slower than that) is won't work on these controller.

avisconti replied:
ack

tbursztyka commented on drivers/sensor/lsm6dsl/lsm6dsl_spi.c:
```diff
@@ -0,0 +1,106 @@
+static struct spi_config lsm6dsl_spi_conf = {
+	.frequency = 16000000,
+	.operation = SPI_OP_MODE_MASTER | SPI_MODE_CPOL |
+	SPI_MODE_CPHA | SPI_WORD_SET(8) | SPI_LINES_SINGLE,
```
indentation seems a bit wrong. Check your editor if it can apply for you the linux code style, so it will do the indentation properly etc...

avisconti replied:
some of the coding style thing are already set in my vim.
I'll change this manually.

tbursztyka commented on drivers/sensor/lsm6dsl/lsm6dsl_spi.c:
```diff
@@ -0,0 +1,106 @@
+	};
+
+	if (len > 64)
+		return -EIO;
```
we enforce using {} always, even for one liners statement on if/for/while...

Apply that comment where relevant

avisconti replied:
OK, I read the coding style exceptions to Linux now.
Thx!

tbursztyka commented on drivers/sensor/lsm6dsl/lsm6dsl_spi.c:
```diff
@@ -0,0 +1,106 @@
+	if (spi_transceive(spi_cfg, &tx_buf, 1, &rx_buf, 1))
+		return -EIO;
+
+	memcpy(value, &buffer_rx[1], len);
```
You are misusing the spi API here, the buffering was made especially to avoid memcopies.

you need to rework your rx_buf:

struct spi_buf rx_buf[] = {
{
.buf = NULL,
.len = 1
},
{
.buf = value,
.len = 1
},
};

That should do it. Changes spi_transceive parameters accordingly.

avisconti replied:
think I got it. I'll give a try.

avisconti commented:
Pushed a new version

tbursztyka commented on drivers/sensor/lsm6dsl/lsm6dsl.c:
```diff
@@ -20,11 +21,8 @@
-	if (i2c_reg_update_byte(data->i2c_master, config->i2c_slave_addr,
-				LSM6DSL_REG_CTRL3_C,
-				LSM6DSL_MASK_CTRL3_C_BOOT,
+	if (data->hw_tf->update_reg(data, LSM6DSL_REG_CTRL3_C, LSM6DSL_MASK_CTRL3_C_BOOT,
```
you will need to re-align subsequent lines of parameters. 

On function signature/call, if parameters are broken into different lines, the alignment is made against previous '(' + 1 space.

Apply that everywhere.

Beware of the 80 chars limit btw.

avisconti replied:
I'm pushing again. Now every line is <= than 80 char. And I took care to modify all function calls (at least the one that I changed) with parameter aligned to last '(' + 1 space.

galak commented:
You need a spi version of the dts binding dts/bindings/sensor/st,lsm6dsl.yaml

tbursztyka commented:
@galak this is not possible yet to get a proper spi dts bindings. I'll come with these.

avisconti commented:
@galak @tbursztyka 
So, what are next steps now?

tbursztyka commented:
seems like it still has 80 chars limit issues: https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/11481/4/tests

If you want you may add the spi_cs_control stuff. It's not that much work: see configure_spi() in drivers/ieee802154/ieee802154_cc1200.c for instance

avisconti commented:
> seems like it still has 80 chars limit issues:
> https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/11481/4/tests

I realized now that I have tabstop=4 in my vim. I'll change it to 8.
That;s why I add those errors....

> If you want you may add the spi_cs_control stuff. It's not that much work: 
> see configure_spi() in drivers/ieee802154/ieee802154_cc1200.c for instance

OK, I'll take a look at it, but I won't be able to test it on any platform. I'll implement it and maybe 
you can help me in reviewing it.

avisconti commented:
Pushed again.
This time I ran checkpatch and fixed all warnings. Sorry if I didn't do it before.
I added SPI_CS as per your suggestion. It should work, but I do not have any platform to test it.

avisconti commented:
@bogdan-davidoaia 
Can you possibly take a look to this patch as well?

Thx!

erwango commented on drivers/sensor/lsm6dsl/Kconfig:
```diff
@@ -8,41 +8,93 @@
 	hex
 	prompt "LSM6DSL I2C address"
-	depends on LSM6DSL
+	depends on LSM6DSL && LSM6DSL_I2C
```
Is LSM6DSL dependency actually needed here? (same question everywhere)

avisconti replied:
Maybe not, but it is there everywhere...

erwango commented on drivers/sensor/lsm6dsl/Kconfig:
```diff
@@ -8,41 +8,93 @@
-	default I2C_0_NAME
+	default "SPI_2"
 	help
 	  Specify the device name of the I2C master device to which LSM6DSL is
```
the master

avisconti replied:
ack

erwango commented on drivers/sensor/lsm6dsl/lsm6dsl.c:
```diff
@@ -364,66 +352,74 @@ static int lsm6dsl_init_chip(struct device *dev)
-	if (!data->i2c_master) {
+	data->comm_master = device_get_binding(config->comm_master_dev_name);
+	if (!data->comm_master) {
 		SYS_LOG_DBG("i2c master not found: %s",
```
master not found

erwango commented on drivers/sensor/lsm6dsl/lsm6dsl.h:
```diff
@@ -689,12 +688,23 @@
 struct lsm6dsl_data {
-	struct device *i2c_master;
+	struct device *comm_master;
+	struct lsm6dsl_config *cfg;
```
Is that required to put config into data?
as it is already provided to driver thanks to DEVICE_AND_API_INIT macro

avisconti replied:
It is not. I'll remove it.

erwango commented on drivers/sensor/lsm6dsl/Kconfig:
```diff
@@ -8,41 +8,93 @@
+	  Enable driver for LSM6DSL accelerometer and gyroscope
 	  sensor.
 
-if !HAS_DTS_I2C_DEVICE
```
If not using dst SPI bindings yet, at least you should not remove it for I2C.

avisconti replied:
Yes, probably I misunderstood the purpose of this flag. I leave it.

avisconti commented on drivers/sensor/lsm6dsl/CMakeLists.txt:
```diff
@@ -1 +1,3 @@
-zephyr_sources_ifdef(CONFIG_LSM6DSL lsm6dsl.c)
+zephyr_sources_ifdef(CONFIG_LSM6DSL         lsm6dsl.c)
+zephyr_sources_ifdef(CONFIG_LSM6DSL_SPI     lsm6dsl_spi.c)
+zephyr_sources_ifdef(CONFIG_LSM6DSL_I2C     lsm6dsl_i2c)
```
thx. Strangely enough was picking up the file correctly....

galak commented:
> @galak this is not possible yet to get a proper spi dts bindings. I'll come with these.
> 
> wrong button

why is that not possible?

avisconti commented:
@erwango 
Did I miss any point highlighted by you? I think I implemented all the requested change.
Moreover I also added the TRIGGER part to lsm6dsl code. Should I wait until this PR will be closed or should I put it inside this PR? I tend to prefer the former.

erwango commented on drivers/sensor/lsm6dsl/Kconfig:
```diff
@@ -8,11 +8,11 @@
+	  Enable driver for LSM6DSL accelerometer and gyroscope
 	  sensor.
 
 if !HAS_DTS_I2C_DEVICE
```
I think you should move this below "choice/endchoice" section and put it around I2C config for now

avisconti replied:
ooops... yes correct!

erwango commented on drivers/sensor/lsm6dsl/Kconfig:
```diff
@@ -22,26 +22,82 @@ config LSM6DSL_DEV_NAME
+	  This option is mandatory to set which GPIO pin to use in order
+	  to actually emulate the SPI CS.
+
+config LSM6DSL_MASTER_DEV_NAME
```
You cannot rename this config as is.
"CONFIG_LSM6DSL_I2C_MASTER_DEV_NAME" is generated from device tree (when device tree option is available)
So you should keep CONFIG_LSM6DSL_I2C_MASTER_DEV_NAME and add  CONFIG_LSM6DSL_SPI_MASTER_DEV_NAME

avisconti replied:
ok, got it. I had now to take a deeper look to all this stuff looking at another board. Still I'm not getting all the points, but some of the puzzle pieces are starting to be attached together.

erwango commented on drivers/sensor/lsm6dsl/Kconfig:
```diff
@@ -8,24 +8,41 @@
 
-if !HAS_DTS_I2C_DEVICE
-
 config LSM6DSL_DEV_NAME
```
So, actually this one should be under !HAS_DTS_I2C_DEVICE.
Can be put in depend on maybe

erwango commented:
recheck

erwango commented:
Created #6584 to handle bus configuration depending on bus type.

erwango commented on drivers/sensor/lsm6dsl/Kconfig:
```diff
@@ -8,13 +8,30 @@
+choice
+	prompt "Interface to AP"
+	depends on LSM6DSL
+	default LSM6DSL_SPI
```
Default should be _I2C, esle we'll have to rework all other boards that are using I2C today (by default) and which don't specify it

erwango commented on drivers/sensor/lsm6dsl/Kconfig:
```diff
@@ -8,13 +8,30 @@
+
+endchoice
+
 if !HAS_DTS_I2C_DEVICE
```
We need to adapt here.
Actually driver could be compiled with this flag just because it is used for other drivers and then LSM6DSL_DEV_NAME is no more visible.

So LSM6DSL_DEV_NAME should be outside if !HAS_DTS_I2C_DEVICE
and should have following dependency:
```depends on (LSM6DSL_I2C && !HAS_DTS_I2C_DEVICE) || LSM6DSL_SPI```

avisconti replied:
:(

avisconti commented:
@erwango @tbursztyka 
I re-pushed again.
Now, in addition to Erwan's observations, I also introduced a new transfer routine,
write_data, so I kindly ask you to check the spi and i2c implementation.

struct lsm6dsl_transfer_function {
>       int (*read_data)(struct lsm6dsl_data *data, u8_t reg_addr,
>       >       >       u8_t *value, u8_t len);
>       int (*write_data)(struct lsm6dsl_data *data, u8_t reg_addr,
>       >       >       u8_t *value, u8_t len);
>       int (*read_reg)(struct lsm6dsl_data *data, u8_t reg_addr,
>       >       >       u8_t *value);
>       int (*update_reg)(struct lsm6dsl_data *data, u8_t reg_addr,
>       >       >       u8_t value, u8_t mask);
};

Thanks,
Armando

PS
I'm exausted!! :(

erwango commented on drivers/sensor/lsm6dsl/lsm6dsl.h:
```diff
@@ -689,12 +688,24 @@
+
+struct lsm6dsl_transfer_function {
+	int (*read_data)(struct lsm6dsl_data *data, u8_t reg_addr,
+			u8_t *value, u8_t len);
```
Align on upper line struct, apply to following lines

avisconti replied:
OK, I'll change all the things highlighted.
Did you check the (*write_data) implementation especially the SPI one?
I tested it on ArgonKey and should be ok, but would be nice to give a loom at it.

erwango replied:
Tested ok on disco_l475_iot1

erwango commented on drivers/sensor/lsm6dsl/lsm6dsl_i2c.c:
```diff
@@ -0,0 +1,58 @@
+static u16_t lsm6dsl_i2c_slave_addr = CONFIG_LSM6DSL_I2C_ADDR;
+
+static int lsm6dsl_i2c_read_data(struct lsm6dsl_data *data, u8_t reg_addr,
+			u8_t *value, u8_t len)
```
Align on upper line "struct". Apply everywhere (this file and following one)

erwango commented on drivers/sensor/lsm6dsl/lsm6dsl_i2c.c:
```diff
@@ -0,0 +1,58 @@
+			u8_t *value, u8_t len)
+{
+	return i2c_burst_read(data->comm_master, lsm6dsl_i2c_slave_addr,
+			       reg_addr, value, len);
```
Align on upper line "data", apply everywhere

erwango commented on drivers/sensor/lsm6dsl/lsm6dsl_spi.c:
```diff
@@ -0,0 +1,157 @@
+static int lsm6dsl_raw_read(struct spi_config *spi_cfg, u8_t reg_addr,
+			u8_t *value, u8_t len)
+{
+	u8_t buffer_tx[2] = { reg_addr | LSM6DSL_SPI_READ, 0 };
```
Remove space after "{" and before "}"

erwango commented on drivers/sensor/lsm6dsl/lsm6dsl_spi.c:
```diff
@@ -0,0 +1,157 @@
+static int lsm6dsl_raw_write(struct spi_config *spi_cfg, u8_t reg_addr,
+			u8_t *value, u8_t len)
+{
+	u8_t buffer_tx[1] = { reg_addr & ~LSM6DSL_SPI_READ};
```
remove space after "{"

erwango reviewed (changes requested):
Some formatting fixes and we're good!
Cheers up!!

avisconti commented:
@tbursztyka @erwango 
Guys, I had to repush again because I made a mistake in the update_reg() function signature.
Basically I just inverted mask and value position:

int (*update_reg)(struct lsm6dsl_data *data, u8_t reg_addr, u8_t mask, u8_t value);

I have retested on both ArgonKey (SPI) and nucleo_f411re (I2C).
